inference: fix bad effects for recursion

vtjnash opened the pull request:
Effects are not converged, so they need to always be correct, and not a bestguess, even during recursion. This could be refined, but we don't really need to, and it might be unnecessarily costly to do so.

vtjnash committed:
inference: fix bad effects for recursion

Effects are not converged, so they need to always be correct, and not a
bestguess, even during recursion.

Keno commented:
Fixes #51090

aviatesk commented on base/compiler/typeinfer.jl:
```diff
@@ -849,16 +856,18 @@ function typeinf_edge(interp::AbstractInterpreter, method::Method, @nospecialize
-        return EdgeCallResult(frame.bestguess, edge, frame.ipo_effects) # effects are adjusted already within `finish`
+        isinferred = is_inferred(frame)
+        edge = isinferred ? mi : nothing
+        effects = isinferred ? frame.ipo_effects : adjust_effects(Effects(), method) # effects are adjusted already within `finish` for ipo_effects
```
`Effects()` might be too wide and it causes external `AbsInt` with overlay method table to see inference regression for `typejoin(a, b)` again and leads to the test failure in test/compiler/AbstractInterpreter.jl (xref: #48097).

Maybe we should employ a similar technology as what we do for type convergence?

vtjnash replied:
Except that these are often not converged for types either, since know that is expensive and usually not beneficial, and which is how I encountered this. I think you might need to go back and a special case for it that slows down inference when it is using overlay tables.

It seems related to this regression too, since it is now trying abstract inference on typejoin, even though it should not:
```
julia> length(which(typejoin,(Any,Any)).specializations)
1 # v1.6
90 # v1.9
163 # v1.10
```
I added a commit to fix that, using your recent work on `@nospecializeinfer`

vtjnash committed:
improve effects of typejoin and simplify implementation of Tuple _compute_eltype

  - Repair definition of Core._foldable_meta
  - Handle Vararg better in eltype
  - Mark has_free_typevars ccall as total
  - Mark tailjoin foldable explicitly also, since it calls typejoin recursively
  - Handle non-constructable Tuple{:a,2} and Tuple{T} types
  - Prepare special code for inlining (improves generic effects too)
  - Only optimize typejoin in concrete eval, not needlessly generating
    specializations otherwise on specific types

The compiler seems better able to deal with this version of
_compute_eltype, and it does not seem that necessary to form the IdSet.

vtjnash committed:
inference: teach tfunc that typeof tuples must have fixed lengths

aviatesk commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -2041,13 +2049,14 @@ function abstract_call_known(interp::AbstractInterpreter, @nospecialize(f),
         end
     elseif la == 3 && istopfunction(f, :!==)
         # mark !== as exactly a negated call to ===
+        call = abstract_call_gf_by_type(interp, f, ArgInfo(fargs, Any[Const(f), Any, Any]), si, Tuple{typeof(!==),Any,Any}, sv, max_methods)
```
```suggestion
        call = abstract_call_gf_by_type(interp, f, ArgInfo(fargs, Any[Const(f), Any, Any]), si, Tuple{typeof(f),Any,Any}, sv, max_methods)
```
and better to move this to the line just above `return call` because this isn't needed when `rty` is known to be `Const|Conditional`?

aviatesk commented on base/compiler/tfuncs.jl:
```diff
@@ -731,8 +731,13 @@ function typeof_concrete_vararg(t::DataType)
-            if isdefined(p, :T) && !isdefined(p, :N) && isconcretetype(p.T)
-                return Type{Tuple{t.parameters[1:np-1]..., Vararg{p.T, N}}} where N
+            if isdefined(p, :T) && isconcretetype(p.T)
+                nothing
```
```suggestion
```

aviatesk commented on base/compiler/typeinfer.jl:
```diff
@@ -850,16 +857,18 @@ function typeinf_edge(interp::AbstractInterpreter, method::Method, @nospecialize
-        return EdgeCallResult(frame.bestguess, edge, frame.ipo_effects) # effects are adjusted already within `finish`
+        isinferred = is_inferred(frame)
+        edge = isinferred ? mi : nothing
+        effects = isinferred ? frame.ipo_effects : adjust_effects(Effects(), method) # effects are adjusted already within `finish` for ipo_effects
```
I came up with this fix for `typejoin` accuracy for external `AbstractInterpreter` with overlay table:
```suggestion
        if isinferred
            effects = frame.ipo_effects # effects are adjusted already within `finish`
        else
            # if this method isn't overlayed, we don't need to taint `:nonoverlayed`
            effects = is_nonoverlayed(method) ? EFFECTS_UNKNOWN : Effects()
            effects = adjust_effects(effects, method)
        end
```

But I just realized this isn't correct, because we can't guarantee that any overlay methods aren't involved within a recursive call graph..
And now I think this may be more problematic that I thought, because now we basically make external `AbstractInterpreter` with overlay table unable to perform concrete eval on any recursive methods.

I think there are two options:
1. make inference keep track of a set of methods that appear within a call graph and check if there are any overlay methods here, which sounds a bit complicated but at the same time may be useful for the other analyses too
2. move forward #51080 and annotate `:nonoverlayed typejoin`, which seems awkward honestly, because it is obvious that `typejoin` is not overlayed

vtjnash replied:
`typejoin` is probably considered to be `:total`, which would possibly already work. The only reason this method is a problem is that it is annotated `:terminates`, since otherwise inference couldn't derive that information itself anyways.

maleadt commented:
As this breaks GPU codegen, see https://github.com/JuliaGPU/GPUCompiler.jl/issues/506, I would propose not backporting this anymore so late in the 1.10 release process (cc @KristofferC).

vtjnash commented:
Effects must be correct, otherwise we might eg tab-complete something that crashes the whole process or other bad side effects like that

maleadt commented:
Sure, but the fix is flawed, or at least it regresses GPU compilation: https://github.com/JuliaGPU/GPUCompiler.jl/issues/506. So unless you want to trade one issue for another, I'd rather we not break the GPU stack between 1.10-rc1 and the release.
